Kernel/Devices/SerialDevice.cpp

```diff
@@ -140,4 +92,212 @@ u8 SerialDevice::get_line_status() const
     return m_base_addr.offset(5).in<u8>();
 }
 
+ssize_t SerialDevice::on_tty_write(const UserOrKernelBuffer& buffer, ssize_t size)
+{
+    if (!size)
+        return 0;
+
+    if (!(get_line_status() & EmptyTransmitterHoldingRegister))
+        return 0;
+
+    auto result = buffer.read_buffered<128>(size, [&](u8 const* data, size_t data_size) {
+        for (size_t i = 0; i < data_size; i++)
+            put_char(data[i]);
+        return data_size;
+    });
+    if (result.is_error())
+        return result.error();
+    return (ssize_t)result.value();
+}
+
+void SerialDevice::echo(u8 ch)
+{
+    put_char(ch);
+}
+
+void SerialDevice::discard_pending_input()
+{
+    set_fifo_control(ClearReceiveFIFO);
+}
+
+void SerialDevice::discard_pending_output()
+{
+    set_fifo_control(ClearTransmitFIFO);
+}
+
+void SerialDevice::wait_until_pending_output_completes()
+{
+    while ((get_line_status() & EmptyDataHoldingRegisters) == 0)
+        ;
+}
+
+int SerialDevice::change_baud(speed_t in_baud, speed_t out_baud)
+{
+    if (in_baud != out_baud) {
+        dbgln("{}: Input and output speed must be the same", m_tty_name);
+        return -ENOTSUP;
+    }
+    auto maybe_baud = serial_baud_from_termios(in_baud);
+    if (maybe_baud.has_value()) {
+        set_baud(maybe_baud.value());
+    } else {
+        dbgln("{}: Tried to set unsupported baud speed", m_tty_name);
+        return -ENOTSUP;
+    }
+    return 0;
```

Comment:
this can be more neatly written in the early return pattern, e.g.
```suggestion
    if (!maybe_baud.has_value()) {
        dbgln("{}: Tried to set unsupported baud speed", m_tty_name);
        return -ENOTSUP;
    }
    set_baud(maybe_baud.value());
    return 0;
```